File: JavaScript/102. Binary Tree Level Order Traversal.js
// noinspection JSUnusedGlobalSymbols

/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
const levelOrder = (root) => {
  if (!root) return []

  const nodes = [root]
  const tree = []

  while (nodes[0]) {
    const length = nodes.length
    const row = []

    for (let i = 0; i < length; i++) {
      const node = nodes.shift()
      if (node) {
        row.push(node.val)

        if (node.left) nodes.push(node.left)

        if (node.right) nodes.push(node.right)
      }
    }
    tree.push(row)
  }
  return tree
}

/**
 * Definition for a binary tree node.
 */
const TreeNode = function (val = 0, left = null, right = null) {
  this.val = val
  this.left = left
  this.right = right
}

File: Go/102. Binary Tree Level Order Traversal.go
package main

//goland:noinspection GoUnusedFunction
func levelOrder(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}

	nodes := []*TreeNode{root}
	var tree [][]int

	for len(nodes) > 0 {
		length := len(nodes)
		var row []int

		for i := 0; i < length; i++ {
			node := nodes[0]
			nodes = nodes[1:]

			if node != nil {
				row = append(row, node.Val)

				if node.Left != nil {
					nodes = append(nodes, node.Left)
				}

				if node.Right != nil {
					nodes = append(nodes, node.Right)
				}
			}
		}

		tree = append(tree, row)

	}

	return tree
}
